rasa/shared/core/domain.py

```diff
@@ -930,10 +930,13 @@ def _get_user_sub_state(
 
         # filter entities based on intent config
         # sub_state will be transformed to frozenset therefore we need to
-        # convert the list to the tuple
+        # convert the set to the tuple
         # sub_state is transformed to frozenset because we will later hash it
         # for deduplication
-        entities = tuple(self._get_featurized_entities(latest_message))
+        entities = tuple(
+            self._get_featurized_entities(latest_message)
```

Comment:
Nearly there but not entirely. 

This still fails because `self.latest_message` is just a pointer to the event.
```python
def test_tobi():
    tracker = DialogueStateTracker.from_events(
        "Vova",
        evts=[
            ActionExecuted(ACTION_LISTEN_NAME),
            UserUttered("hi", intent={"name": "greet"}),
            DefinePrevUserUtteredFeaturization(True),
            DefinePrevUserUtteredEntities(
                entities=[{"entity": "entity1", "value": "value1"}]
            ),
        ],
    )

    user = list(tracker.events)[1]
    assert isinstance(user, UserUttered)

    # Fails
    assert not user.entities
```

Reply:
it should fail

Reply:
we want user uttered to have entities

Reply:
but this means that when we store the tracker, we store the updated version of `UserUttered` which would change history.

Reply:
that's fine

Reply:
featurizers take care of it

Reply:
In my opinion I don't think it's fine. It means that the `TrackerStore` is no longer a 1:1 presentation of what happened. This break the events' property of immutability.

Reply:
It depends when you write events to the tracker store. I think we already discussed it, and I thought you fixed it. We should only persist `UserUttered` event after policy prediction happened and all the policy events are applied

Reply:
There is no need to rely on a fragile thing such as the time of persistence. 

> We should only persist UserUttered event after policy prediction happened and all the policy events are applied

There is no need for that as we have the two additional / new events. We will just replay events and come back to the current state. When we discussed the timing of persistence, it was when we considered if we need the additional events after all.

Reply:
yeah, but we use them differently

Reply:
>  It means that the TrackerStore is no longer a 1:1 presentation of what happened.

which is good, since we effectively made `UserUttered` a dynamic event

Reply:
This is good for the ML models, but it's not good for the system itself. Breaking the immutability property is opening the box of the pandora in my opinion.

Reply:
I saw the other comment, I see your point